bigquery/storage/managedwriter/retry.go

```diff
@@ -18,35 +18,73 @@ import (
 	"context"
 	"errors"
 	"io"
+	"strings"
 	"time"
 
 	"github.com/googleapis/gax-go/v2"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
+var (
+	defaultAppendRetries = 3
+)
+
+func newDefaultRetryer() *defaultRetryer {
+	return &defaultRetryer{
+		bigBo: gax.Backoff{
+			Initial:    2 * time.Second,
+			Multiplier: 5,
+			Max:        5 * time.Minute,
+		},
+	}
+}
+
 type defaultRetryer struct {
-	bo gax.Backoff
+	bo    gax.Backoff
+	bigBo gax.Backoff // For more aggressive backoff, such as throughput quota
 }
 
 func (r *defaultRetryer) Retry(err error) (pause time.Duration, shouldRetry bool) {
-	// TODO: refine this logic in a subsequent PR, there's some service-specific
-	// retry predicates in addition to statuscode-based.
+	// This predicate evaluates errors for both enqueuing and reconnection.
+	// See RetryAppend for retry that bounds attempts to a fixed number.
 	s, ok := status.FromError(err)
 	if !ok {
 		// Treat context errors as non-retriable.
 		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
 			return r.bo.Pause(), false
 		}
+		// EOF can happen in the case of connection close.
+		if errors.Is(err, io.EOF) {
+			return r.bo.Pause(), true
+		}
 		// Any other non-status based errors treated as retryable.
 		return r.bo.Pause(), true
 	}
 	switch s.Code() {
-	case codes.Unavailable:
+	case codes.Aborted,
+		codes.Canceled,
+		codes.DeadlineExceeded,
+		codes.Internal,
+		codes.Unavailable:
 		return r.bo.Pause(), true
-	default:
-		return r.bo.Pause(), false
+	case codes.ResourceExhausted:
+		if strings.HasPrefix(s.Message(), "Exceeds 'AppendRows throughput' quota") {
+			// Note: internal b/246031522 opened to give this a structured error
+			// and avoid string parsing.  Should be a QuotaFailure or similar.
+			return r.bigBo.Pause(), true // more aggressive backoff
+		}
```

Comment:
In case is a `codes.ResourceExhausted`, but not the more specific quota errors, shouldn't we retry too ?

Reply:
Abstractly, this PR represents a widening of cases where we'll retry.  We can widen further in the future if we find there's more opportunities to improve the status quo.

I expect some of the other existing quotas like open connection count will be less problematic as we work towards multiplexing support, which should net yield far less open connections.